### PR TITLE
Bootstrap workspace discovery: fix infinite loop, pom model cache

### LINKED ISSUE #10549: Infinite loop resolving workspace
If you take the reproducer from https://github.com/quarkusio/quarkus/issues/5879 and update the quarkus version to the latest version you will get an infinite loop when running mvn quarkus:dev. 

This project has an odd structure, where the parent is not the top level project.

The issue is in this code here: https://github.com/quarkusio/quarkus/blob/35272d70b5c077748f6e136d6d7be166b575637f/independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java#L146

Reading the model will correctly follow the parent references to get to the correct parent, but then it looks in the top level directory again, and will restart the cycle. I am not really sure what the correct fix is here, because I don't know why that logic was added. 

Also this code seems inefficient, if we have parsed the model shouldn't we be saving it somewhere so we don't need to re-parse it when loading the workspace?

### PR DESCRIPTION
**Pull request by aloubyansky**
Fixes #10549

### CODE REVIEW

File: independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java [hunk 1]
```diff
@@ -30,6 +31,124 @@
...
+                model = loadAndCache(pomFile);
+            }
+            return model;
+        }
+
+        private Model loadAndCache(Path pomFile) throws BootstrapMavenException {
+            final Model model = readModel(pomFile);
+            cachedModels.put(pomFile.getParent(), model);
+            return model;
+        }
```
Comment (gastaldi) [inline independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java:69]: This snippet can be replaced with: 

```java
        private Model model(Path pomFile) throws BootstrapMavenException {
            return cachedModels.computeIfEmpty(pomFile.getParent(), parent->readModel(pomFile));
        }
```

Comment (aloubyansky) [reply independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java:69]: computeIfAbsent()? But then I'd have to catch the exception and re-throw a runtime one.

Comment (gastaldi) [reply independent-projects/bootstrap/maven-resolver/src/main/java/io/quarkus/bootstrap/resolver/maven/workspace/LocalProject.java:69]: Ah damn checked exceptions, you're right :(

**Review by gastaldi — APPROVED** [review]

**Review by stuartwdouglas — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
@aloubyansky do you think this is safe enough to backport? I am kinda on the fence about it, the inifite loop is nasty but it has only come up once that we have seen so I don't think it is super common. The reduced parsing of pom files might be nice though.

**Comment by aloubyansky** [comment]
I have pretty much the same feeling. I think I would have backported this primarily to fix the bug.

**Pull request merged**